### PR TITLE
Test 3d texture sub-source uploads from ImageData

### PR DESCRIPTION
**Pull request by kainino0x**
(And replace the tests for obsolete entry points.)

I pretty much rewrote the whole test, sorry for the huge diff. I think it runs reasonably fast despite being pretty thorough.

Verified in Chromium: https://codereview.chromium.org/2500433003

### CODE REVIEW

File: sdk/tests/js/tests/tex-image-and-sub-image-3d-with-image-data.js [hunk 1]
```diff
@@ -256,19 +257,33 @@ function generateTest(internalFormat, pixelFormat, pixelType, prologue, resource
...
         }
+        // XXX: remove this stuff
+        //if (bindingTarget == gl.TEXTURE_2D_ARRAY) {
+        //    return;
+        //}
+        //if (true) { const sub = true;
+        //    if (true) { const flipY = false;
+        //        if (true) { const premul = false;
+        //            if (true) { var rect = rects[1];
+        //                if (true) { const rTexCoord = 0;
```
Comment (kenrussell) [inline sdk/tests/js/tests/tex-image-and-sub-image-3d-with-image-data.js]: Remove this stuff. :)

File: sdk/tests/js/tests/tex-image-and-sub-image-3d-with-image-data.js [hunk 2]
```diff
@@ -256,19 +257,33 @@ function generateTest(internalFormat, pixelFormat, pixelType, prologue, resource
...
         for (const sub of [false, true]) {
             for (const flipY of [false, true]) {
                 for (const premul of [false, true]) {
-                    for (var irect = 0; irect < rects.length; irect++) {
+                    for (let irect = 0; irect < rects.length; irect++) {
```
Comment (kenrussell) [inline sdk/tests/js/tests/tex-image-and-sub-image-3d-with-image-data.js]: Is "let" supported well in all browsers now?

File: sdk/tests/js/tests/tex-image-and-sub-image-3d-with-image-data.js [hunk 3]
```diff
@@ -205,49 +213,42 @@ function generateTest(internalFormat, pixelFormat, pixelType, prologue, resource
...
             [0, 0, 2, 4],
             [2, 0, 2, 4],
-            [0, 2, 4, 2],
         ];
-        // NOTE: depth and unpackImageHeight are not actually simulated, so
-        // new test cases with different values may not actually work.
-        var depth = 2;
-        var unpackImageHeight = 2;
-        var dbg = false;
+        var dbg = true;
```
Comment (kenrussell) [inline sdk/tests/js/tests/tex-image-and-sub-image-3d-with-image-data.js]: Do we want to leave "dbg" true?

**Comment by kainino0x** [comment]
Ready now, PTAL

**Comment by kainino0x** [comment]
@zhenyao @kenrussell

File: sdk/tests/js/tests/tex-image-and-sub-image-3d-with-image-data.js (hunk 3 above)
Comment (kainino0x) [inline sdk/tests/js/tests/tex-image-and-sub-image-3d-with-image-data.js]: It's false now.

File: sdk/tests/js/tests/tex-image-and-sub-image-3d-with-image-data.js (hunk 1 above)
Comment (kainino0x) [inline sdk/tests/js/tests/tex-image-and-sub-image-3d-with-image-data.js]: Done

File: sdk/tests/js/tests/tex-image-and-sub-image-3d-with-image-data.js (hunk 2 above)
Comment (kainino0x) [inline sdk/tests/js/tests/tex-image-and-sub-image-3d-with-image-data.js]: Based on MDN (plus a quick test in Safari), yes, let and const are well supported.

**Comment by kenrussell** [comment]
Superb work Kai. These tests look excellent. Thank you for your persistence getting them fully working.

**Pull request merged**